ec/ecresource/organizationresource/update.go

```diff
@@ -0,0 +1,291 @@
+		new.Organization,
+	)
+
+	addDeployment, removeDeployment := diffDeploymentRoleAssignments(
+		old.Deployment,
+		new.Deployment,
+	)
+
+	var addProject, removeProject *models.ProjectRoleAssignments
+	if old.Project != nil && new.Project != nil {
+		addProject, removeProject = diffProjectRoleAssignments(
+			*old.Project,
+			*new.Project,
+		)
+	} else if old.Project == nil && new.Project != nil {
+		addProject, removeProject = new.Project, nil
+	} else if old.Project != nil && new.Project == nil {
+		addProject, removeProject = nil, old.Project
+	}
+
+	add := models.RoleAssignments{
+		Organization: addOrganization,
+		Deployment:   addDeployment,
+		Project:      addProject,
+	}
+	remove := models.RoleAssignments{
+		Organization: removeOrganization,
+		Deployment:   removeDeployment,
+		Project:      removeProject,
+	}
+
+	return add, remove
+}
+
+func diffOrganizationRoleAssignments(old, new []*models.OrganizationRoleAssignment) ([]*models.OrganizationRoleAssignment, []*models.OrganizationRoleAssignment) {
+	getKey := func(ra models.OrganizationRoleAssignment) string {
+		return *ra.RoleID
+	}
+	add := difference(new, old, getKey)
+	remove := difference(old, new, getKey)
+	return add, remove
+}
+
+func diffDeploymentRoleAssignments(old, new []*models.DeploymentRoleAssignment) ([]*models.DeploymentRoleAssignment, []*models.DeploymentRoleAssignment) {
+	getKey := func(ra models.DeploymentRoleAssignment) string {
+		var all bool
+		if ra.All != nil {
+			all = *ra.All
+		}
+		sort.Strings(ra.DeploymentIds)
+		return fmt.Sprintf("%s-%t-%s", *ra.RoleID, all, strings.Join(ra.DeploymentIds, ","))
+	}
+	add := difference(new, old, getKey)
+	remove := difference(old, new, getKey)
+	return add, remove
+}
+
+func diffProjectRoleAssignments(old, new models.ProjectRoleAssignments) (*models.ProjectRoleAssignments, *models.ProjectRoleAssignments) {
+	getKey := func(ra models.ProjectRoleAssignment) string {
+		var all bool
+		if ra.All != nil {
+			all = *ra.All
+		}
+		sort.Strings(ra.ProjectIds)
+		return fmt.Sprintf("%s-%t-%s", *ra.RoleID, all, strings.Join(ra.ProjectIds, ","))
+	}
+
+	addElasticsearch := difference(new.Elasticsearch, old.Elasticsearch, getKey)
+	removeElasticsearch := difference(old.Elasticsearch, new.Elasticsearch, getKey)
+
+	addObservability := difference(new.Observability, old.Observability, getKey)
+	removeObservability := difference(old.Observability, new.Observability, getKey)
+
+	addSecurity := difference(new.Security, old.Security, getKey)
+	removeSecurity := difference(old.Security, new.Security, getKey)
+
+	add := models.ProjectRoleAssignments{
+		Elasticsearch: addElasticsearch,
+		Observability: addObservability,
+		Security:      addSecurity,
+	}
+	remove := models.ProjectRoleAssignments{
+		Elasticsearch: removeElasticsearch,
+		Observability: removeObservability,
+		Security:      removeSecurity,
+	}
+	return &add, &remove
+}
+
+func difference[T interface{}](a, b []*T, getKey func(T) string) []*T {
```

Comment:
IMO we need unit test for this func (as for other funcs (at least internal ones) in this package).

Reply:
Does it make sense to use something like https://github.com/hashicorp/go-set/blob/main/hashset.go instead?

Reply:
Yeah I will check if that simplifies the code

Reply:
Yeah, I suppose we can proceed with the difference function, just cover it by tests.

Reply:
:+1: I'm adding some tests for the difference function